Add wrapper for core functionality

MaHaWo opened the pull request:
closes #31 

- [x] add prototype
- [x] add `change_analyzer` function 
- [x] check if multiprocessing/threading is really necessary and change implementation accordingly
- [x] add unit tests
- [x] make sonarcloud shut up

MaHaWo commented on src/iSparrow/preprocessor_base.py:
```diff
@@ -4,6 +4,10 @@
 import audioread
 
 
+class AudioFormatError(Exception):
```
Needed because it is thrown below.

MaHaWo commented on src/iSparrow/preprocessor_base.py:
```diff
@@ -49,14 +53,13 @@ def read_audio_data(self, path: str) -> np.array:
         except audioread.exceptions.NoBackendError as e:
-            print(e)
-            raise AudioFormatError("Audio format could not be opened.")
+            raise AudioFormatError("Audio format could not be opened.") from e
```
... better exception handling...

MaHaWo commented on src/iSparrow/utils.py:
```diff
@@ -145,3 +148,47 @@ def load_module(alias: str, path: str):
+    return getattr(module, name)
+
+
+def wait_for_file_completion(file_path: str, polling_interval=1) -> bool:
```
Function that continually checks if a file still changes its size. Only way I could find with to make sure that a file is not read before it was written without making assumptions about the program that writes the file.

MaHaWo commented on src/iSparrow/sparrow_recording.py:
```diff
@@ -194,8 +191,14 @@ def read_audio_data(self):
+        # README: wait until the file to be opened does not change in size anymore.
+        # only working way in which we do not have to make assumptions about the
+        # system that actually writes the file?
+        utils.wait_for_file_completion(str(self.path), polling_interval=1)
```
this blocks until the file is ready. checks every 1 second.

MaHaWo commented on models/birdnet_default/preprocessor.py:
```diff
@@ -43,7 +43,7 @@ def process_audio_data(self, rawdata: np.ndarray) -> list:
         Returns:
             list: chunked audio data
         """
-        print("process audio data custom")
```
these outputs where wrongly named

MaHaWo commented on src/iSparrow/utils.py:
```diff
@@ -8,14 +8,40 @@
     pass
 
 
+def read_yaml(path: str):
```
Has been moved here from the set_up script, because it's needed in various places.

MaHaWo commented on src/iSparrow/utils.py:
```diff
@@ -131,17 +157,63 @@ def load_model_from_file_torch(path: str, _):
-    spec = importlib.util.spec_from_file_location(alias, path)
-    module = importlib.util.module_from_spec(spec)
-    spec.loader.exec_module(module)
+    try:
```
this needed a better exception handling... was missing before and was overlooked.

MaHaWo commented on src/iSparrow/utils.py:
```diff
@@ -131,17 +157,63 @@ def load_model_from_file_torch(path: str, _):
     return module
+
+
+def load_name_from_module(module_name: str, file_path: str, name: str):
```
this makes things a little more compact when dynamically loading modules

MaHaWo commented on tests/conftest.py:
```diff
@@ -5,19 +5,38 @@
 
 # add a fixture with session scope that emulates the result of a later to-be-implemented-install-routine
-@pytest.fixture(scope="session", autouse=True)
+@pytest.fixture(scope="module", autouse=True)
```
... limit interaction between tests.

MaHaWo commented on tests/fixtures/watcher_fixtures.py:
```diff
@@ -0,0 +1,168 @@
+from copy import deepcopy
+
+
+class WatchFixture:
```
This class concentrates all helper function that are needed in the tests and all the data that they need. Helper funcitons mostly live here because sonarcloud complains about code repetition if I just spell them out in the tests themselves.

MaHaWo commented on src/iSparrow/sparrow_watcher.py:
```diff
@@ -0,0 +1,527 @@
+import multiprocessing
+
+
+class AnalysisEventHandler(FileSystemEventHandler):
```
custom handler class that uses the sparrowrecording to analyze incoming audio files.

MaHaWo commented on src/iSparrow/sparrow_watcher.py:
```diff
@@ -0,0 +1,527 @@
+
+
+def watchertask(watcher):
+    """
```
this lives outside of the watcher class because of multiprocessing.

MaHaWo commented on src/iSparrow/sparrow_watcher.py:
```diff
@@ -0,0 +1,527 @@
+            watcher: (SparrowWatcher): Watcher this Handler is used with
+        """
+        self.pattern = watcher.pattern
+        self.recording = watcher.set_up_recording()
```
it uses the setup capabilities of the watcher, but all the modules for the models etc is loaded  only here, which allows them to live in the watcher process and not the main process. (the latter approach created problems on macos).

MaHaWo commented on src/iSparrow/sparrow_watcher.py:
```diff
@@ -0,0 +1,527 @@
+        self.species_predictor_config = species_predictor_config
+
+        # restart process to make changes take effect
+        self.restart()
```
actual exchange only happens once the process is restarted.

MaHaWo commented on src/iSparrow/sparrow_watcher.py:
```diff
@@ -0,0 +1,527 @@
+
+        self.watcher_process = None
+
+        self.may_do_work = multiprocessing.Event()
```
these 2 events are here to indicate when the analyzer is done with the current file (`is_done_analyzign`, needed to not create corrupt analysis files) and for stopping it from picking up a new one when it should be restarted, or to allow it to be paused (`may_do_work`).

MaHaWo commented on src/iSparrow/sparrow_watcher.py:
```diff
@@ -0,0 +1,527 @@
+
+        for filename in self.input.iterdir():
+
+            condition = (
```
make sure that the file that's currently under analysis is not taken into account upon cleanup if this is run concurrently.

MaHaWo commented on src/iSparrow/sparrow_watcher.py:
```diff
@@ -0,0 +1,527 @@
+
+        if self.reanalyze_on_cleanup:
+            recording = self.set_up_recording()
+
```
clean_up does not use the recording of the daemon process.

MaHaWo commented:
@iulusoy so, this thing would be ready now for review. However, given it's size, let me know if I should look into splitting it up. It got bigger and bigger during development, and would have warranted a meta issue in hindsight....

MaHaWo commented:
@iulusoy ok, so this is ready for review in principle. I say "in principle", because it became very big during development, to the point where it would have warranted its own meta issue. Therefore, if you judge it to be too big, I can try and split it up into multiple PRs. The feature as such is quite self-contained though, I would say.

MaHaWo commented on notebooks/usage.ipynb:
```diff
@@ -43,36 +43,70 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Creating iSparrow folders and downloading data... \n",
```
.. [it should though, shouldn´t it?](https://github.com/ssciwr/iSparrow/blob/master/.pre-commit-config.yaml)

MaHaWo commented on requirements-dev.txt:
```diff
@@ -3,4 +3,5 @@ myst-parser
-pre-commit
\ No newline at end of file
+pre-commit
+ipykernel
```
hm, the example notebook shouldn´t need anything that iSparrow itself doesn´t? what other components would that be?

iulusoy replied:
If you want to run the notebook outside of your VSCode (VSCode may have notebook extension installed, but not necessarily needs to run the notebook in vscode).

MaHaWo replied:
... opted for removal of notebooks for now. Can be re-added when doing documentation I think.

MaHaWo commented on notebooks/usage.ipynb:
```diff
@@ -43,36 +43,70 @@
-    "import iSparrow.sparrow_model_base as spm\n",
-    "import iSparrow.sparrow_recording as spr\n",
-    "import iSparrow.preprocessor_base as spb\n",
+    "from iSparrow import PreprocessorBase\n",
```
There's an explicit remark above the code lines that the package must be installed first from the repo. Then it should work. (later it will be installed via pip anyway, right? )

MaHaWo commented on src/iSparrow/sparrow_watcher.py:
```diff
@@ -0,0 +1,515 @@
+        outdir: str,
+        model_dir: str,
+        model_name: str,
+        preprocessor_config: dict = {},
```
done.

MaHaWo commented on notebooks/usage.ipynb:
```diff
@@ -43,36 +43,70 @@
+   "source": [
+    "import sys\n",
+    "\n",
+    "sys.path.append(str(Path.home() / Path(\"Development\") / \"iSparrow\"))"
```
replaced it with a relative include of the package. Since this is only used for the setup method, maybe the solution would be to have a setup module in the src folder? Would add this in a separate PR that does the packaging and install stuff?

MaHaWo commented on tests/fixtures/watcher_fixtures.py:
```diff
@@ -0,0 +1,170 @@
+            )
+
+            wait_for_file_completion(Path(data) / Path(f"example_{i}.wav"))
+
```
the stuff below is mostly there because sonarcloud complains about repetition if I put it into the test code itself

MaHaWo commented:
@iulusoy ok, I think this is ready for a second round. 
- I removed the `clean_up` and `change_analyzer` method and (will) put them into separate branches to  make the size a little more manageable. 
- I left in the multiprocessing approach though, as preparation as it were. I removed the notebook because it's just an internal documentation thing. I also added python 3.9 and python 3.12 for testing, because the raspberry uses 3.9 currently, and 3.12 is just there because it's the newest one at the moment, I use it on macos and windows for instance. 
- Other than that I added additional dependencies (ffmpeg) that silences a silent error message from an audio library that librosa uses.
